sbi/inference/snpe/snpe_c.py

```diff
@@ -231,5 +370,354 @@ def _log_prob_proposal_posterior(
+        logit_factors = logits_p_rep + logits_d_rep
+
+        # Compute sqrt(det()/(det()*det()))
+        covariances_p = torch.inverse(precisions_p)
+        covariances_d = torch.inverse(precisions_d)
+
+        logdet_covariances_pp = torch.logdet(covariances_pp)
+        logdet_covariances_p = torch.logdet(covariances_p)
+        logdet_covariances_d = torch.logdet(covariances_d)
+
+        # Repeat the proposal and density estimator terms such that there are LK terms.
+        # Same trick as has been used above.
+        logdet_covariances_p_rep = logdet_covariances_p.repeat_interleave(
+            num_comps_d, dim=1
+        )
+        logdet_covariances_d_rep = logdet_covariances_d.repeat(1, num_comps_p)
+
+        sqrt_det_ratio = 0.5 * (
+            logdet_covariances_pp
+            - (logdet_covariances_p_rep + logdet_covariances_d_rep)
+        )
+
+        # Compute for proposal, density estimator, and proposal posterior:
+        # mu_i.T * P_i * mu_i
+        exponent_p = _batched_vmv(precisions_p, means_p)
+        exponent_d = _batched_vmv(precisions_d, means_d)
+        exponent_pp = _batched_vmv(precisions_pp, means_pp)
+
+        # Extend proposal and density estimator exponents to get LK terms.
+        exponent_p_rep = exponent_p.repeat_interleave(num_comps_d, dim=1)
+        exponent_d_rep = exponent_d.repeat(1, num_comps_p)
+        exponent = -0.5 * (exponent_p_rep + exponent_d_rep - exponent_pp)
+
+        logits_pp = logit_factors + sqrt_det_ratio + exponent
+
+        return logits_pp
+
+    def _maybe_z_score_theta(self, theta: Tensor) -> Tensor:
+        """Return potentially standardized theta if z-scoring was requested."""
+
+        if self.z_score_theta:
+            theta, _ = self._posterior.net._transform(theta)
+
+        return theta
+
+
+def _batched_vmv(matrix: Tensor, vector: Tensor) -> Tensor:
+    """
+    Returns (vector.T * matrix * vector).
+
+    Doing this with einsum() allows for vector and matrix to be batched.
+
+    Args:
+        matrix: Matrix of shape
+            (batch_dim, num_components, parameter_dim, parameter_dim).
+        vector: Vector of shape (batch_dim, num_components, parameter_dim).
+
+    Returns:
+        Product (vector.T * matrix * vector) of shape (batch_dim, num_components).
+    """
+    return torch.einsum("bci, bci -> bc", vector, _batched_mv(matrix, vector))
+
+
+def _batched_mv(matrix: Tensor, vector: Tensor) -> Tensor:
+    """
+    Returns (matrix * vector).
+
+    Doing this with einsum() allows for vector and matrix to be batched.
+
+    Args:
+        matrix: Matrix of shape
+            (batch_dim, num_components, parameter_dim, parameter_dim).
+        vector: Vector of shape (batch_dim, num_components, parameter_dim).
+
+    Returns:
+        Product (matrix * vector) of shape (batch_dim, num_components, parameter_dim).
+    """
+    return torch.einsum("bcij,bcj -> bci", matrix, vector)
+
+
+def _mog_log_prob(
```

Comment:
and this should probably go into `pyknos.mdn` at some point, but fine for me to leave it here until we code a new MDN in `pyknos`.

Reply:
I made a pyknos issue here: https://github.com/mackelab/pyknos/issues/16